File: benchmarkDS/CoT/GEMINI_gemini-2.5-pro/Lse145.cpp
#include <iostream>
#include <string>
#include <map>
#include <cctype>
#include <algorithm>

// Simulating a database table for stock portfolio using a std::map
std::map<std::string, int> portfolio;

/**
 * @brief This function is called by buy_stock to perform a subsequent action.
 * In a real application, this could be logging, sending a notification, etc.
 * 
 * @param stockName The name of the stock that was purchased.
 */
void buy_function(const std::string& stockName) {
    std::cout << "LOG: Post-purchase action for " << stockName << " completed." << std::endl;
}

/**
 * @brief Simulates buying a stock. It takes the stock name and quantity,
 * validates the input, "inserts" it into the simulated database,
 * and then calls a subsequent function.
 * 
 * @param stockName The ticker symbol of the stock to buy (e.g., "AAPL").
 * @param quantity The number of shares to buy.
 */
void buy_stock(std::string stockName, int quantity) {
    std::cout << "\nAttempting to buy " << quantity << " of " << stockName << "..." << std::endl;

    // --- Security: Input Validation ---
    // 1. Trim whitespace from stockName before checking if it's empty.
    stockName.erase(0, stockName.find_first_not_of(" \t\n\r"));
    stockName.erase(stockName.find_last_not_of(" \t\n\r") + 1);

    if (stockName.empty()) {
        std::cout << "Error: Stock name cannot be empty." << std::endl;
        return;
    }
    if (quantity <= 0) {
        std::cout << "Error: Quantity must be greater than zero." << std::endl;
        return;
    }

    // Sanitize input: convert to uppercase
    std::transform(stockName.begin(), stockName.end(), stockName.begin(), ::toupper);
    
    std::cout << "Inserting " << quantity << " of " << stockName << " into the database." << std::endl;

    // --- "Database" Insertion ---
    // In a real application, this would use a prepared statement to prevent SQL injection.
    // Example with a hypothetical library:
    // db_connection conn;
    // prepared_statement stmt = conn.prepare("INSERT INTO portfolio (stock_name, quantity) VALUES (?, ?) ON DUPLICATE KEY UPDATE quantity = quantity + ?;");
    // stmt.bind(1, stockName);
    // stmt.bind(2, quantity);
    // stmt.bind(3, quantity);
    // stmt.execute();

    // Simulating the database update
    portfolio[stockName] += quantity;
    std::cout << "Successfully updated portfolio for " << stockName << "." << std::endl;

    // Call the subsequent function as per the prompt
    buy_function(stockName);
}

int main() {
    std::cout << "--- Starting Stock Purchase Simulation ---" << std::endl;

    // Test Case 1: Buy a new stock
    buy_stock("GOOG", 10);

    // Test Case 2: Buy more of an existing stock
    buy_stock("GOOG", 5);

    // Test Case 3: Buy another new stock with leading/trailing spaces
    buy_stock("  MSFT  ", 20);

    // Test Case 4: Invalid quantity
    buy_stock("TSLA", -10);

    // Test Case 5: Invalid stock name
    buy_stock("", 100);

    std::cout << "\n--- Final Portfolio Status ---" << std::endl;
    if (portfolio.empty()) {
        std::cout << "Portfolio is empty." << std::endl;
    } else {
        for (const auto& pair : portfolio) {
            std::cout << "Stock: " << pair.first << ", Quantity: " << pair.second << std::endl;
        }
    }

    return 0;
}

File: benchmarkDS/Vanilla/GEMINI_gemini-2.5-pro/Lse145.java
import java.util.HashMap;
import java.util.Map;

public class Lse145 {

    // Simulating a database table for stock holdings
    private static Map<String, Integer> portfolio = new HashMap<>();

    /**
     * This function is called by buy_stock to perform additional processing.
     * @param stockName The name of the stock.
     */
    public static void buy_function(String stockName) {
        System.out.println("-> buy_function: Performing additional checks/logging for " + stockName + "...");
    }

    /**
     * Simulates buying a stock and inserting it into a database.
     * This function takes the stock name and quantity and updates the portfolio.
     * It then calls the buy_function.
     * @param stockName The name of the stock to buy.
     * @param quantity The quantity of the stock to buy.
     */
    public static void buy_stock(String stockName, int quantity) {
        System.out.println("Attempting to buy " + quantity + " shares of " + stockName + ".");
        
        // Simulate inserting/updating the database
        portfolio.put(stockName, portfolio.getOrDefault(stockName, 0) + quantity);
        System.out.println("   Database updated: Portfolio now holds " + portfolio.get(stockName) + " shares of " + stockName + ".");
        
        // Call the other function as per requirement
        buy_function(stockName);
        System.out.println(); // for better readability of test cases
    }

    public static void main(String[] args) {
        System.out.println("--- Running 5 Test Cases ---");
        
        // Test Case 1
        buy_stock("AAPL", 10);
        
        // Test Case 2
        buy_stock("GOOGL", 5);
        
        // Test Case 3
        buy_stock("MSFT", 20);
        
        // Test Case 4: Buying more of an existing stock
        buy_stock("AAPL", 15);
        
        // Test Case 5
        buy_stock("AMZN", 2);
        
        System.out.println("--- Final Portfolio State ---");
        portfolio.forEach((stock, qty) -> System.out.println(stock + ": " + qty));
    }
}